### PR TITLE
Flat arrays pave the ways

### PR DESCRIPTION
**Pull request by rjpolackwich**
This PR does three main things:

Accomplishes some important backend optimizations, including: 

-Database optimization. Arrays are now flat, meaning that indexing or partitioning into groups can be dynamic or set flexibly, as well as improves access performance and makes every part of IO easier to integrate with. We now write out index tables as available.

-HTTP client optimization. There is now one client class. This class can be used and configured to do just about everything we might want it to do, eg optimized data streaming with batch-configurable io callbacks (ie, build h5 database _in batch_ while consuming a stream data by data...). 
--Callbacks can be configured with great flexibility, eg, running in place, delegating to a designated io threadpool, designating to a delegated cpu-threadpool, running with a lock, and any combination thereof. This means complicated functions can run in the background _while_ we consume data. This means that fire-and-forget operations can be run, meaning that pushing data across a network from a client could be accomplished (send data out to Samasource, for instance, or whatever). It also means that we can accumulate dataset statistics as data comes in on the fly (cpu threadpool), or write arbitrary data out to any number of databases ("split this dataset into two new datasets: planes over land, and planes over water").
--Notably, filtering is supported as well, and can be written with functions that data, image labels or any user-provided framework (eg, "give me only samples with both cars and planes" or, more advanced but perhaps more useful, , "give me data until it fits _this distribution among features_", which means that I can define a distribution across my data for each feature so that I can, eg, specify a negative/positive ratio _for each feature within the dataset_ on the fly, meaning we can know exactly what our model sees for each feature from the start).
--Single entry input. Give the client a source iterable, tell it when you want it to stop, add more sources as you consume based on a callback you provide...

Accomplishes some important architecture optimizations, including:

--A "accessor-focused" philosophy. Let me explain what I mean by that: We have many different attributes that we tend to need to pass to many different kinds of objects at any particular time. Some of these attributes are more important in particular situations, eg, partitions need to be defined when doing ML training, but aren't really necessary during dataset QA, etc. New objects may require new attributes, and some attributes may need to do certain things when they're set. We deal with a lot of this at the class instantiation level, but this becomes heavy and not very flexible in development (what does specifying a count mean if we want to iterate over multiple datasets until a condition is met?), as well as confusing and cumbersome at the user level (what does dtype mean, what is image size, etc). 

The approach taken here to make using our accessor classes more easily is to focus on the accessor itself, and it's structure, to answer questions like: What input arguments are unique and required for any use of this accessor? How does it do iteration under the hood? What special and unique methods are supplied by this accessor? And so on. 

Technical: To deal with whatever myriad of input arguments regarding data and classes and stuff like that, our classes delegate `**kwargs` down to the base class instantiation, which sets up all the properties as specified by the class via something called `vprops`. Now we can write alllllll of our properties out by themselves in one place, where we can mix them together and do other special things, and then put them on our accessors with a decorator.  

TL;DR: Accessor code is more readable, easier to develop with/from, easier to combine, and so on. We accomplish this by separating property defining sets from the iterators themselves in the codebase.

With this approach, we can define what our accessors can do with functions defined  _outside of the accessor_. 

--An entry-level typing-system `mltypes`. We have some fundamental kinds of things we deal with all the time, right? classification things, segmentation things, label things, image things, sample things, combinations thereof, whatever. We've historically figured out what we need to do everywhere by doing stuff like "if mltype=="classification", do this, otherwise, do this, but if not that, do this thing that's not implemented yet. With types, we can make comparisons and take correct actions according to some fundamental descriptor, and write helper functions and checker fucntions for things we do a lot. Now we can do things like look at all of our available types, or immediately get the correct action by doing things like `mltypes.types_match`, and build new types from these fundamental types in python, eg: we could support 6 sample types  between [image, image] and the conbimations of [image, label] types. Feel free to do `isinstance(obj, mltypes.BinaryClassificationType)` kinda stuff too, now.

We can extend the typing concept to define functions that can help us accomplish one of the original goals: immediate model training over datasets where the first epoch is network-based and the following are disk-based. That in and of itself is a fundamental win for this library, imo.

Defines some new APIS, including

--A pretty flexible transform api. `thing.train.images.transforms.register(user_function)`. Kind of a mouthful, could be shortened. You can chain functions, look at all your functions defined, give them names, specify the ordering, delete them, etc. Same for labels too. We can make hooks off this that let us do visual comparisons between input and the output of the chain, too. Do you think that would be nice? I think people would like that so that they can write their own things and make sure they're doing what they want them to do. And we could write cool predefined ones for samples, like @martham93 has done, register them on the fly, update them, whatever! `thing.train.transforms.register([pv.Flip, pv.Flop, pv.Blur, pv.Whatever, AlsoMycustomFunctionToo, pv.DisplayComparison])`, and now `thing.train[0]` will return the result of the chain and also display some fancy comparison orvalidation or whatever. 

--A filtering api for on the fly data access. "Give me all the cars in veda that match..." The way this works is in place (implemented in the way the client works) but I haven't actually looked at where the api should exist. The structure for the api does exist, however, so it's just a design decision from here.

That's a basic general overview of the work that's been here, and hopefully provides insight into the motivation for how it all works together. This is obviously a non-standard PR that introduces a lot of new concepts, many of them internal, but it should hopefully make building out new functionality a lot easier going forward for devs, as well as making the things we can build more intuitive at the interface level.

Look for a notebook soon that demonstrates some of this functionality.  Feedback encouraged below.

[Notebook here](https://github.com/DigitalGlobe/veda-docs/blob/db8909bc6f3b803e9ae29a7dfb4adab8d36d0302/PyVeda/pveda-pr-55.ipynb)

### CODE REVIEW

**Comment by rjpolackwich** [comment]
All is working, tests passing with couple small exceptions that can be fixed afterwards, will add the updated notebook associated with this to the docs, explains how to use transform plugins etc.
